File: Two_Pointers/Python/15_three_sum.py
# RUNTIME: 1298 ms, faster than 33.25% of python3 Submissions. 
# MEMORY USAGE: 20.02 MB, less than 99.45% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Define Target as 0
# 2. Set left to i + 1 and right to len(nums) - 1
# 3. if sum of left and right is less than target, we need to increase left
# 4. if sum of left and right is greater than target, we need to decrease right
# 5. continue until solution is found or left is greater than right

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        target = 0
        nums.sort()
        output = set()
        result = []
        for i in range(len(nums)):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                sum = nums[i] + nums[left] + nums[right]
                if sum == target:
                    output.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1
                elif sum < target:
                    left += 1
                else:
                    right -= 1
        result = list(output)
        return result

File: Two_Pointers/C++/15_three_sum.cpp
/*
* RUNTIME: 786 ms - Beats 36.31% of submissions
* MEMORY - 190.52 MB - Beats 32.46% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
using namespace std;


/* Thoughtprocess:
* 1. initailize a left and right pointer
* 2. create a set to store the unique triplets
* 3. Initiate a loop to iterate through the vector
* 4. Use two pointers to browse through the vector
* 5. If the sum of the three numbers is equal to the target, add the triplet to the set
* 6. If the sum is less than the target, move the left pointer to the right
* 7. If the sum is greater than the target, move the right pointer to the left
* 8. Repeat untill no possibilities are left
* 9. Return the set as a vector
* 
*/

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int target = 0;
        set <vector<int>> s;
        int left, right;
        int sum;
        sort(nums.begin(), nums.end());

        for (int i = 0; i < nums.size(); i++)
        {
            left = i + 1;
            right = nums.size() - 1;
            while (left < right)
            {
                sum = nums[i] + nums[left] + nums[right];
                if ( sum == target)
                {
                    s.insert({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                }
                else if (sum < target)
                    left++;
                else
                    right--;
            }
        }
        return vector<vector<int>> (s.begin(), s.end());
    }
};